emily/lambda/src/operations/chainstate.rs

```diff
@@ -2,50 +2,148 @@
+        common::deserialize_request::<models::CreateChainstateRequestContent>(body)?;
+
+    let chainstate_entry = models::CreateChainstateResponseContent {
+        block_hash: request.block_hash.clone(),
+        block_height: request.block_height,
+    };
+
+    let mut item: HashMap<String, AttributeValue> = HashMap::new();
+    item.insert("BlockHeight".to_string(), AttributeValue::N(request.block_height.to_string()));
+    item.insert("BlockHash".to_string(), AttributeValue::S(request.block_hash));
+
+    context.dynamodb_client
+        .put_item()
+        .table_name(context.settings.chainstate_table_name.clone())
+        .set_item(Some(item))
+        .send()
+        .await
+        // TODO:
+        // Make error provide less internal info.
+        .map_err(|e|
+            EmilyApiError::InternalService(
+                format!("Error occurred for table {:?} {:?}", context.settings.chainstate_table_name.clone(), e)
+            )
+        )?;
+
+    common::package_response(chainstate_entry, 201)
 }
 
 /// Handles retrieval of the current chainstate
-pub fn handle_get_chainstate(
-    _path_parameters:  HashMap<String, String>
+pub async fn handle_get_chainstate(
+    _path_parameters: HashMap<String, String>,
+    context: &LambdaContext,
 ) -> Result<common::SimpleApiResponse, errors::EmilyApiError> {
-    // path_parameters.keys()
-    Ok(models::GetChainstateResponseContent {
-        block_hash: "Tortoise".to_string(),
-        block_height: 42.0,
-    })
-    .and_then(|response| {
-        common::package_response(response, 200)
-    })
+    // // path_parameters.keys()
+    let height: &String = _path_parameters.get("height").unwrap();
+    let query_output = context.dynamodb_client
+        .query()
+        .table_name(context.settings.chainstate_table_name.clone())
+        .key_condition_expression("#h = :height")
+        .expression_attribute_names("#h", "BlockHeight")
+        .expression_attribute_values(":height", AttributeValue::N(height.clone()))
+        .send()
+        .await
+        // TODO:
+        // Make error provide less internal info.
+        .map_err(|e|
+            EmilyApiError::InternalService(
+                format!("Error occurred for table {:?} {:?}", context.settings.chainstate_table_name.clone(), e)
+            )
+        )?;
+
+    // Change behavior depending on how many entries were found.
+    match query_output.items().len() {
+
+        // No blocks found.
+        0 => Err(EmilyApiError::NotFound(
+            format!("No bock hash found for height {}", height)
+        )),
+
+        // A single block found; happy path.
+        1 => {
+            // Extract.
+            let item = query_output.items().first().unwrap();
+
+            // TODO:
+            // extract fields more safely.
+            let block_hash = item.get("BlockHash").unwrap().as_s().unwrap();
+            let block_height = item.get("BlockHeight").unwrap().as_n().unwrap();
```

Comment:
If we expect these conversions to happen often, we could break out some of the boilerplate by implementing `TryFrom<HashMap<String, AttributeValue>>` for `GetChainResponseContent`. Another alternative would be to create a custom trait for this conversion (`trait FromAttributeValueMap`) - which we might need since the `GetChainResponseContent` type seems to be external and the orphan rule would prohibit us from implementing an external trait on an external type.

Reply:
I see that you're using `TryFrom` a lot for the types in `resources/deposits.rs`. That's great, and we could define wrapper types here to be able to implement TryFrom for the external types as well.

```
struct GetChainstateResponseContent(models::GetChainResponseContent)

impl TryFrom<...> for GetChainstateResponseContent {...}

etc.
```

Reply:
Also if you use `serde_dynamo` as @djordon suggested that might be even better than the custom conversion definitions.